dynamic_instrumentation/src/point/mod.rs

```diff
@@ -23,25 +24,87 @@ pub struct InstrumentationPoint<'tcx> {
     pub args: Vec<InstrumentationArg<'tcx>>,
     pub is_cleanup: bool,
     pub after_call: bool,
+    pub instrumentation_priority: usize,
     pub metadata: EventMetadata,
 }
 
+/// Gathers a set of all address-taken locals in a function
+/// body.
+///
+/// The set may include address-taken arguments.
+pub struct CheckAddressTakenLocals<'a, 'tcx: 'a> {
+    /// The set of address-taken locals.
+    pub address_taken: IndexSet<Local>,
+    tcx: TyCtxt<'tcx>,
+    body: &'a Body<'tcx>,
+}
+
+impl<'a, 'tcx: 'a> CheckAddressTakenLocals<'a, 'tcx> {
+    pub fn new(tcx: TyCtxt<'tcx>, body: &'a Body<'tcx>) -> Self {
+        Self {
+            address_taken: IndexSet::new(),
+            tcx,
+            body,
+        }
+    }
+
+    pub fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+}
+
+/// Rewrites all address-taken locals in terms of their underlying
+/// address.
+///
+/// For example, if `_x` is an address-taken local and `_y` is the local
+/// storing the address of `_x`, the statement `_z = _x` will be
+/// rewritten as `_z = (*_y)`.
+pub struct RewriteAddressTakenLocals<'tcx> {
+    /// The set of address-taken locals.
+    pub address_taken: IndexSet<Local>,
+    /// A mapping from the address-taken local to the local
+    /// storing the former's address.
+    pub local_substitute: IndexMap<Local, Local>,
+    tcx: TyCtxt<'tcx>,
+}
+
+impl<'tcx> RewriteAddressTakenLocals<'tcx> {
+    pub fn new(tcx: TyCtxt<'tcx>, address_taken: IndexSet<Local>) -> Self {
+        Self {
+            address_taken,
+            local_substitute: IndexMap::new(),
+            tcx,
+        }
+    }
+
+    pub fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+}
+
 pub struct CollectInstrumentationPoints<'a, 'tcx: 'a> {
     tcx: TyCtxt<'tcx>,
     hooks: Hooks<'tcx>,
-    body: &'a Body<'tcx>,
-    instrumentation_points: Vec<InstrumentationPoint<'tcx>>,
+    pub body: &'a Body<'tcx>,
+    pub instrumentation_points: Vec<InstrumentationPoint<'tcx>>,
```

Comment:
What foreign context now consumes these?

Reply:
`CollectInstrumentationPoints::visit_assign` on line 389:

```rust
let num_statements = self.body.basic_blocks()[location.block].statements.len();
```

Reply:
also line 564 in `visit_body`:

```rust
        if self.instrumentation_points.is_empty() {
            return;
        }
```

Reply:
Ah, it's because we implement the visitor trait in terms of the struct's internals in a separate module from where we define the visitor struct itself.